### ISSUE TITLE
[tune] Multiple population base training (PBT) problems with TF2

### ISSUE THREAD

**Issue by timost1234** [ray-project/ray#11936]
Hi, I am using the population based training in ray tune with TensorFlow 2.3, there are 3 questions I have encountered.

1) During the training progress, every time the perturbation happens, the iteration was also reset to 0. I don't think this was working correctly. I have tried to modify the `_init_ self.iteration`  in `TuneReporterCallback` to 25 and `initial_epoch=50` in `model.fit`, however, the "training_iteration" in result.json still starts from 1. I am really curious about how ray tune find the current iteration as well as how to fix this problem.

2)  My training accuracy also drops a lot when the perturbations happens. The example code in PyTorch also performs like this, so I don't consider this as a bug in my code. But why does this happen? It doesn't seem like the accuracy will drop according to the PBT paper.

3) Many of the hyperparameters went out of the range that I set in `hyperparam_mutations`, I am wondering if any body had encountered this situation?

Thank a lot for your help.

**Comment by amogkam** [comment]
Hi @timost1234 thanks for raising this issue! All of these seem very odd and could potentially be bugs on our end. Do you have a script I could run to reproduce these issues? I can then investigate further. Also what version of Ray are you using?

**Comment by timost1234** [comment]
Hello @amogkam.  Thanks for your prompt reply.

I am using ray version 1.1.0.dev0. Please find the full script as well as some output in [this colab file](https://colab.research.google.com/drive/1VMn77tU3DQCg--006tAe-kEIGGiIqBVx?usp=sharing). I also state the problems again in the file. Hope these will be helpful. 

BTW, I really like `ray tune`. Is there any document explaining how this library was implemented (like how do they interact with other libraries)?

**Comment by amogkam** [comment]
Hi @timost1234 just briefly looking at your Colab notebook, I think one issue is that checkpoints are not properly being used in your training function. See this guide for an in-depth example, https://docs.ray.io/en/latest/tune/tutorials/tune-advanced-tutorial.html#function-api-with-population-based-training, but your `tune_model` function needs to accept an additional `checkpoint_dir` arg to load and save the model. This could be causing problems 1 and 2.

**Comment by timost1234** [comment]
Hi @amogkam I exposed the `checkpoint_dir` per the tutorial, however, it didn't seem to work. [Here](https://colab.research.google.com/drive/1Hv8R7kHhRJdvZ-pSR9nVA2BVTxwGGQAw?usp=sharing) are the revised script. Thanks again for your help.

**Comment by amogkam** [comment]
Ah got it, so since you are using Keras, you should use the `TuneReportCheckpointCallback` (https://docs.ray.io/en/latest/tune/api_docs/integration.html?highlight=keras#ray.tune.integration.keras.TuneReportCheckpointCallback) which will automatically report _and_ checkpoint for you. This is the only callback that's need and will handle saving and registering checkpoints with Tune. But you should keep the part where you load the checkpoint if `checkpoint_dir` is not None. Give it a try and let me know if it works. If not, I can run through the notebook tomorrow to see what's going on.

**Comment by timost1234** [comment]
Hello @amogkam The `TuneReportCheckpointCallback` do solve the problem. 
It now left only problem 3. Do you have any idea about why the HPs were going out of the range set in `hyperparam_mutations`?

**Comment by amogkam** [comment]
Hi @timost1234 sorry for the late reply! For the `hyperparam_mutations` going out of bounds, this is to be expected. From your notebook it looks like both hyperparam_mutations values are continuous variables, and the way continuous hyperparams are mutated is with a probability of `resample_probability`, the hyperparam gets resampled from the initial distribution, and with probability of `1-resample_probability`,  the hyperparam is multiplied by either 0.8 or 1.2. So it's definitely possible and expected for _continuous_ hyperparameters_ to go out of range. This shouldn't be the case for discrete hyperparameters however. For more information, you can check here https://docs.ray.io/en/master/tune/api_docs/schedulers.html#population-based-training-tune-schedulers-populationbasedtraining or the original PBT paper.

**Comment by Eddiesyn** [comment]
Hi, I'm using pytorch to implement a PBT based search algorithm and i'm facing the same problem here as @timost1234 has stated in No.1 and 2: it seems to me when the perturbation or when the paused trials are starting to run again, it loads the wrong checkpoint. After some experiments i found when the total resources is enough for running all trials in parallel this loading problem is gone and it works just fine. But when the total resources is not enough or when you open the synch mode in PBT, which both means during training, some trials are going to be paused to wait for other tirals to run, the loading problem happens. Is this a bug in function api or is there a work around for pytorch like in Keras?
